### ISSUE TITLE
async variant

### ISSUE THREAD

**Issue by hyusetiawan** [paarthenon/variant#3]
this library at this point has touched all the core parts of our application, we are really enjoy the much stricter type checks with the metadata attached to the different variants, good stuff!

we want to use variant as task definitions, for ex:
```
const TaskExtractMetadata = variant('extract_metadata', async function(file: File) {
// do async stuff
    const stuff1 = await bla();
     return {
         stuff1
     }
})
```

then, we can run the task and get a defined shape of the result that we can exhaustively check as we add more tasks.

The issue is, I don't think variant works with async because it immediately returns with type even though the code above did not throw any typescript error. The code above will run and returns `{type: 'extract_metadata'}` without stuff1

Since the typescript doesnt complain, I may have been using it wrong or this could be a feature request / bug report  :)

**Comment by paarthenon** [comment]
Hello,

Thank you for using the library. I'm so glad it's been working for you. You're right, variant `1.0` didn't support async. I've just gone through and made updates. Try using version `1.1.0`. 

Here's a link to my (passing) test with your code example:

https://github.com/paarthenon/variant/blob/master/src/variant.spec.ts#L115

Please let me know if that works for you.

**Comment by hyusetiawan** [comment]
works as expected, also, this feature delivery is faster than amazon prime <3

**Comment by hyusetiawan** [comment]
I think the async variant return typing is incorrect:
`export type FlowTaskSourceMetadataResult = PromiseValue<ReturnType<typeof FlowTaskSourceMetadataRunner>>`
shows the result correctly, but it's missing the `type` value
not sure if I extract the typing correctly or not, here's how I type that variantList:
```
export const FlowTaskRunner = variantList([FlowTaskSourceMetadataRunner])
export type FlowTaskResult<T extends TypeNames<typeof FlowTaskRunner> = undefined> = VariantOf<typeof FlowTaskRunner, T>
```
and the result of FlowTaskResult is a promise with a type property, I don't think this is intentional

**Comment by paarthenon** [comment]
Hmmm could you give a larger code sample? What version of TypeScript are you using? 

I tried to create an example of what you're describing, but I haven't replicated the issue you're experiencing. I'm on TS 3.8.3 targeting ES2017, for context.

```typescript
export interface FlowTaskSourceMetadata {
    prop1: number;
    prop2: string;
}

export const FlowTaskSourceMetadataRunner = variant('source_metadata', async function FlowTaskSourceMetadataResult(task: FlowTaskSourceMetadata) {
    // does anything meaningful happen here?
    return task;
});

// I don't know PromiseValue but I'm guessing this is what it is.
type PromiseValue<T> = T extends PromiseLike<infer R> ? R : T;
export type FlowTaskSourceMetadataResult = PromiseValue<ReturnType<typeof FlowTaskSourceMetadataRunner>>

export const FlowTaskRunner = variantList([FlowTaskSourceMetadataRunner])
export type FlowTaskResult<T extends TypeNames<typeof FlowTaskRunner> = undefined> = VariantOf<typeof FlowTaskRunner, T>
```

And for me, both `FlowTaskSourceMetadataResult` and `FlowTaskResult` are:

![image](https://user-images.githubusercontent.com/10353928/80645233-d552da80-8a38-11ea-8004-7c782a69707c.png)

**Comment by hyusetiawan** [comment]
typescript version is 3.8.3 and targeting es2018 (i dont think the targeting matters?)
I tried using PromiseValue exactly the one you have with the same result but originally is this:
```
export type PromiseValue<T> = T extends PromiseLike<infer U> ? { 0: PromiseValue<U>; 1: U }[T extends PromiseLike<any> ? 0 : 1] : T

```
here's a longer version:

```
export async function FlowTaskSourceMetadataLogic(task: FlowTaskSourceMetadata) {
	const { source, locale } = task
	const samples = []

	const tableIterator = csvAsyncIterator(source.file, {
		dynamicTyping: false, // all strings, we will figure them out ourselves
		skipEmptyLines: true,
		header: true,
	})

	for await (const row of tableIterator) {
		samples.push(row)
		if (samples.length === source.config.sampleSize) {
			break
		}
	}
	// we have sample
	const tableID = uuid()
	const columns = deduceTableColumnTypes(samples, locale)
	const runID = uuid()

	return {
		metadata: {
			[tableID]: {
				// consistent ordered index, so input can be parameterized, as in, given different file, we get the same parsed table order
				// table name
				name: source.file.name,
				id: tableID, // needs to be globally unique
				runID,
				sourceID: source.id,
				columns,
			},
		},
		samples: {
			[tableID]: {
				rows: samples,
				runID,
			},
		},
	}
}

// export type FlowTaskSourceMetadataResult = PromiseValue<ReturnType<typeof FlowTaskSourceMetadataLogic>>
export const FlowTaskSourceMetadataRunner = variant('source_metadata', FlowTaskSourceMetadataLogic)
export type FlowTaskSourceMetadataResult = PromiseValue<ReturnType<typeof FlowTaskSourceMetadataRunner>>
```

and these are the results:
![Screen Shot 2020-04-29 at 2 24 48 PM](https://user-images.githubusercontent.com/1062920/80648637-5d2ee980-8a25-11ea-94f1-d4dd65fc52d4.png)
![Screen Shot 2020-04-29 at 2 24 57 PM](https://user-images.githubusercontent.com/1062920/80648643-5dc78000-8a25-11ea-93ac-c3e8fdd315b9.png)

**Comment by paarthenon** [comment]
Hey I'm still working on getting an isolated example of this but I noticed something...

> ```typescript
> export const FlowTaskSourceMetadataRunner = variant('source_metadata', FlowTaskSourceMetadataLogic)
> export type FlowTaskSourceMetadataResult = PromiseValue<ReturnType<typeof FlowTaskSourceMetadataRunner>>
> ```
> 
> and these are the results:
> ![Screen Shot 2020-04-29 at 2 24 48 PM](https://user-images.githubusercontent.com/1062920/80648637-5d2ee980-8a25-11ea-94f1-d4dd65fc52d4.png)

I don't think `VariantCreator` should look like that anymore. 

In `1.0` and below, `variant('test', (name: string) => ({name})),` would be of type 

    VariantCreator<'test', [string], {name: string}, 'type'>`

 **but now** it would be of type 

    VariantCreator<'test', (name: string) => {name: string}, 'type'>.

In short, that bracket being there makes me wonder if you somehow have stale definitions since that is the previous form of that type. Have you restarted VS Code or executed `Ctrl+Alt+P` -> TypeScript: Restart TS Server since updating?

**Comment by hyusetiawan** [comment]
I have not, let me try that and report back

**Comment by hyusetiawan** [comment]
oh wow, yeah that is the issue
![Screen Shot 2020-04-29 at 3 57 44 PM](https://user-images.githubusercontent.com/1062920/80654875-2c08e600-8a32-11ea-941d-16c2a6990cb6.png)

who would have thought, the classic IT solution of turning it off and on again applies to typescript too!
